### DISCUSSION TITLE
Disable on filetype

### DISCUSSION THREAD

**Discussion by ls-devs** [mrjones2014/legendary.nvim#386] [Q&A] [answered]
Hi,
I'm currently trying do disable some keymaps for certain filetype and I'm doing this way : 
```lua 
      {
        "<leader>dn",
        function()
          if vim.bo.filetype ~= "cpp" or vim.bo.filetype ~= "c" then
            return "<cmd>Lspsaga diagnostic_jump_next<CR>"
          end
        end,
        description = "LSPSaga Diagnostic Jump Next",
        opts = opts,
      },
```

But this is not working. It seems to be disabled everywhere now.

So my question is : How can I disable some keymaps for certain filetypes ?

**Comment by ls-devs** [comment]
Was doing it wrong.
Right way to do it is : 
```lua
      {
        "<leader>dn",
        function()
          if vim.bo.filetype ~= "cpp" or vim.bo.filetype ~= "c" then
            return require("lspsaga.diagnostic"):goto_next()
          end
        end,
        description = "LSPSaga Diagnostic Jump Next",
        opts = opts,
      },
```

**Reply by mrjones2014** [reply, marked as answer]
Probably the "more correct" way of doing it is to use an autocmd to create the keymap on `filetype`:

```lua
require('legendary').setup({
  autocmds = {
    {
      'Filetype',
      function()
        require('legendary').keymap({
          '<leader>dn',
          require('lspsaga.diagnostic'):goto_next,
          description = 'LSPSaga Diagnostic Jump Next',
          opts = vim.tbl_deep_extend("force", opts, { buffer = 0 }),
        })
      end
      opts = { pattern = { 'c', 'cpp' } },
    },
  }
})
```

Also FYI going to next diagnostic can be done with just `vim.diagnostic.goto_next()`.

**Reply by mrjones2014** [reply]
Alternatively, you could use `legendary.nvim` in a file `after/ftplugin/c.lua` and `after/ftplugin/cpp.lua` that just binds the keymap in `legendary.nvim`.

**Reply by ls-devs** [reply]
@mrjones2014  
Thanks for your reply 
I use `require('lspsaga.diagnostic'):goto_next` because I use this plugin to handle diagnostics.

And I might be wrong but the solution you providing is to enable for filetype c and cpp but not to disable it ?
Is there a way to tell with autocmd that I dont want this command to be mapped for c and cpp files but for all others filetypes ?

**Reply by mrjones2014** [reply]
Oh, my mistake, sorry I misread your code a bit. I thought you wanted it only on c/cpp. I don't think there's a way to _exclude_ specific filetypes in an autocmd pattern, so either you'll have to check it in the callback like you had, or set up an autocmd for the filetypes you _do_ want.

Another alternative approach (and the one I use personally) is to have a set of keymaps that only get assigned in your LSP `on_attach` callback function, and in there, make the keymaps buffer-specific.

```lua
local function on_attach()
  local keymaps = {
    -- the rest of your LSP keymaps here
  }
  if vim.bo.filetype ~= 'c' and vim.bo.filetype ~= 'cpp' then
    table.insert(keymaps, {
          '<leader>dn',
          function()
            require('lspsaga.diagnostic'):goto_next()
          end,
          description = 'LSPSaga Diagnostic Jump Next',
          opts = vim.tbl_deep_extend("force", opts, { buffer = 0 }),
        })
  end
end
```

**Reply by mrjones2014** [reply]
You can either use the `on_attach` param in `lspconfig` or you can use the `LspAttach` autocmd event.
